### PR TITLE
implement Order

### PR DESCRIPTION
**Pull request by Roma1990Melnichuk**

### CODE REVIEW

File: src/main/java/com/bookstore/dto/UpdateOrderStatusRequest.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package com.bookstore.dto;
+
+import lombok.Data;
+
+@Data
+public class UpdateOrderStatusRequest {
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/dto/UpdateOrderStatusRequest.java:7]: Do we need validate request?

File: src/main/java/com/bookstore/dto/OrderRequest.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package com.bookstore.dto;
+
+import lombok.Data;
+
+@Data
+public class OrderRequest {
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/dto/OrderRequest.java:7]: Do we need validate request?

File: src/main/java/com/bookstore/entity/Order.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+@Getter
+@SQLDelete(sql = "Update orders SET is_deleted = TRUE WHERE id = ?")
+@SQLRestriction(value = "is_deleted = FALSE")
+@Table(name = "orders")
+public class Order {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne()
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/Order.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/com/bookstore/entity/Order.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    @Enumerated(EnumType.STRING)
+    private Status status;
+
+    @Column(nullable = false)
+    private BigDecimal total;
+
+    @Column(name = "order_date", nullable = false)
+    private LocalDateTime orderDate;
+
+    @Column(name = "shipping_address", nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/Order.java]:
```suggestion
    @Column(nullable = false)
    private LocalDateTime orderDate;

    @Column(nullable = false)
```

File: src/main/java/com/bookstore/entity/OrderItem.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+@Getter
+@SQLDelete(sql = "Update orderItems SET is_deleted = TRUE WHERE id = ?")
+@SQLRestriction(value = "is_deleted = FALSE")
+@Table(name = "order_items")
+public class OrderItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne()
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/OrderItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/com/bookstore/entity/OrderItem.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    @JoinColumn(name = "order_id", nullable = false)
+    private Order order;
+
+    @ManyToOne()
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/OrderItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/com/bookstore/entity/OrderItem.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    @JoinColumn(name = "book_id", nullable = false)
+    private Book book;
+
+    @Column(nullable = false)
+    private int quantity;
+
+    @Column(nullable = false)
+    private BigDecimal price;
+
+    @Column(name = "is_deleted",nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/OrderItem.java]:
```suggestion
    @Column(nullable = false)
```

File: src/main/java/com/bookstore/mapper/OrderMapper.java [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+import com.bookstore.dto.OrderDto;
+import com.bookstore.dto.OrderItemDto;
+import com.bookstore.entity.Order;
+import com.bookstore.entity.OrderItem;
+import java.util.List;
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+
+@Mapper(componentModel = "spring")
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/mapper/OrderMapper.java]:
```suggestion
@Mapper(componentModel = "spring", uses = OrderItemMapper.class)
```

File: src/main/java/com/bookstore/service/OrderServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+@Service
+public class OrderServiceImpl implements OrderService {
+    private final OrderRepository orderRepository;
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final OrderMapper orderMapper;
+    private final OrderItemMapper orderItemMapper;
+
+    @Transactional
+    @Override
+    public OrderDto orderReturn(OrderRequest request, User user) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/service/OrderServiceImpl.java]:
```suggestion
    public OrderDto createOrder(OrderRequest request, User user) {
```

File: src/main/java/com/bookstore/service/OrderServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        Order order = new Order();
+        order.setShippingAddress(request.getShippingAddress());
+        order.setStatus(Order.Status.PENDING);
+        BigDecimal total = BigDecimal.ZERO;
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUser(user)
+                .orElseThrow(() -> new EntityNotFoundException("Shopping cart not found for user: "
+                        + user.getUsername()));
+        Set<OrderItem> orderItems = new HashSet<>();
+        for (CartItem cartItem : shoppingCart.getCartItems()) {
+            OrderItem orderItem = new OrderItem();
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/service/OrderServiceImpl.java]: let's divide logic
could you use separate methods?

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/com/bookstore/controller/OrderController.java [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+    public List<OrderDto> getOrderHistory(Pageable pageable, Authentication authentication) {
+        User user = (User) authentication.getPrincipal();
+        return orderService.findAll(user, pageable);
+    }
+
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    @Operation(summary = "Update order status",
+            description = "Endpoint for updating the status of an order")
+    @PatchMapping("/{id}")
+    @ResponseStatus(HttpStatus.OK)
```
Comment (liliia-ponomarenko) [inline src/main/java/com/bookstore/controller/OrderController.java]: It’s  default response status

File: src/main/java/com/bookstore/dto/OrderDto.java [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+@Data
+public class OrderDto {
+    private Long id;
+    private Long userId;
+    private List<OrderItemDto> orderItems;
+    private LocalDateTime orderDate;
+    private BigDecimal total;
+    private String status;
+}
+
```
Comment (liliia-ponomarenko) [inline src/main/java/com/bookstore/dto/OrderDto.java]:
```suggestion
```
Only one empty line at the end of the file

File: src/main/java/com/bookstore/dto/OrderRequest.java [hunk 13]
```diff
@@ -0,0 +1,11 @@
+package com.bookstore.dto;
+
+import jakarta.validation.constraints.NotEmpty;
+import lombok.Data;
+
+@Data
+public class OrderRequest {
+    @NotEmpty(message = "Shipping address is required")
```
Comment (liliia-ponomarenko) [inline src/main/java/com/bookstore/dto/OrderRequest.java]: With literals better use NotBlank

File: src/main/java/com/bookstore/dto/UpdateOrderStatusRequest.java [hunk 14]
```diff
@@ -0,0 +1,10 @@
+package com.bookstore.dto;
+
+import jakarta.validation.constraints.NotNull;
+import lombok.Data;
+
+@Data
+public class UpdateOrderStatusRequest {
+    @NotNull(message = "Status is required")
```
Comment (liliia-ponomarenko) [inline src/main/java/com/bookstore/dto/UpdateOrderStatusRequest.java]: the same 
or better use 
```
private Status status;
```

File: src/main/java/com/bookstore/entity/OrderItem.java [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+import java.math.BigDecimal;
+import lombok.Getter;
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
+
+@Entity
+@Setter
+@Getter
+@SQLDelete(sql = "Update orderItems SET is_deleted = TRUE WHERE id = ?")
```
Comment (liliia-ponomarenko) [inline src/main/java/com/bookstore/entity/OrderItem.java]:
```suggestion
@SQLDelete(sql = "UPDATE order_items SET is_deleted = TRUE WHERE id = ?")
```

File: src/main/java/com/bookstore/service/OrderServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+        List<Order> orders = ordersPage.getContent();
+        return orderMapper.toDtoList(orders);
+    }
+
+    @Transactional
+    @Override
+    public OrderDto updateOrderStatus(Long id, Order.Status status) {
+        Order order = orderRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Order not found with id: " + id));
+        order.setStatus(Order.Status.valueOf(status.name()));
```
Comment (liliia-ponomarenko) [inline src/main/java/com/bookstore/service/OrderServiceImpl.java]:
```suggestion
        order.setStatus(status);
```
Why do you complicate it?

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s fix a few comments!

File: src/main/java/com/bookstore/entity/Order.java [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+    @Column(nullable = false)
+    private LocalDateTime orderDate;
+
+    @Column(nullable = false)
+    private String shippingAddress;
+
+    @OneToMany(mappedBy = "order", cascade = CascadeType.ALL, orphanRemoval = true)
+    private Set<OrderItem> orderItems = new HashSet<>();
+
+    @Column(name = "is_deleted",nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/Order.java]:
```suggestion
    @Column(name = "is_deleted", nullable = false)
```

File: src/main/java/com/bookstore/service/OrderServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+    public List<OrderItemDto> getAllOrderItems(Long orderId) {
+        Order order = orderRepository.findById(orderId)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Order not found with id: " + orderId));
+        return orderItemMapper.toDtoList(order.getOrderItems());
+    }
+
+    @Override
+    public OrderItemDto getOrderItem(Long orderId, Long itemId) {
+        Order order = orderRepository.findById(orderId)
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/service/OrderServiceImpl.java]: you could find orderItem by `findByIdAndOrderId`

File: src/main/java/com/bookstore/service/OrderServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+    private ShoppingCart findShoppingCartByUser(User user) {
+        return shoppingCartRepository.findByUser(user)
+                .orElseThrow(() -> new EntityNotFoundException("Shopping cart not found for user: "
+                        + user.getUsername()));
+    }
+
+    private Order buildOrder(OrderRequest request, User user, ShoppingCart shoppingCart) {
+        Order order = new Order();
+        order.setShippingAddress(request.getShippingAddress());
+        order.setStatus(Order.Status.PENDING);
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/service/OrderServiceImpl.java]: minor comment
you can initialise PENDING status and `LocalDateTime.now()` in Order entity

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
almost perfect

File: src/main/java/com/bookstore/controller/OrderController.java [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/orders")
+public class OrderController {
+    private final OrderService orderService;
+
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @Operation(summary = "Place an order", description = "Endpoint for placing an order")
+    @PostMapping
+    public OrderDto placeOrder(@RequestBody OrderRequest request, Authentication authentication) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/controller/OrderController.java]:
```suggestion
    public OrderDto placeOrder(@RequestBody @Valid OrderRequest request, Authentication authentication) {
```

File: src/main/java/com/bookstore/controller/OrderController.java [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+        User user = (User) authentication.getPrincipal();
+        return orderService.findAll(user, pageable);
+    }
+
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    @Operation(summary = "Update order status",
+            description = "Endpoint for updating the status of an order")
+    @PatchMapping("/{id}")
+    public OrderDto updateOrderStatus(@PathVariable Long id,
+                                      @RequestBody UpdateOrderStatusRequest request) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/controller/OrderController.java]:
```suggestion
                                      @RequestBody @Valid UpdateOrderStatusRequest request) {
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
add valid annotation to controller

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request merged**